plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -394,6 +393,157 @@ private S3AsyncClient createS3AsyncClient(AwsCredentialsProvider credentialsProv
         return clientBuilder.build();
     }
 
+    @NotThreadSafe
```

Comment:
It feels like it may be actually better to make this class `ThreadSafe` to avoid global locking in `FileSystemExchangeSource`. Then some of the methods of this class (such as `isBlocked`, `getRetainedSize` and potentially `isFinished`) could be made lock free avoiding unnecessary lock contention.